cmd/restic/cmd_generate.go

```diff
@@ -105,28 +126,45 @@ func runGenerate(opts generateOptions, args []string) error {
 		}
 	}
 
+	err := checkStdoutForSingleShell(opts)
+	if err != nil {
+		return err
+	}
+
 	if opts.BashCompletionFile != "" {
+		if opts.BashCompletionFile == "-" {
```

Comment:
IMO the handling for `-` belongs into `writeBashCompletion` etc. The current changes now introduce two places in which the function to generate the changelog could be called. The current `write*Completion` functions are also lots of copy&paste code.

We could get rid of them using the following function that receives that changelog generator function as parameter:

```go
func writeCompletion(filename string, shell string, generate func(w io.Writer) error) error {
	if stdoutIsTerminal() {
		Verbosef("writing %s completion file to %v\n", shell, filename)
	}
	var outFile io.Writer
	if filename == "-" {
		var err error
		outFile, err = os.Create(filename)
		if err != nil {
			return err
		}
		defer outFile.Close()
	} else {
		outFile = globalOptions.stdout
	}

	return generate(outFile)
}
```

This would then be called using 
```
writeCompletion(opts.BashCompletionFile, "bash", cmdRoot.GenBashCompletion)
```

That would replace all `write*Completion` functions.